treatments.hpp

```diff
@@ -18,80 +18,285 @@
+ */
+template<typename IntegerRaster, typename FloatRaster>
+class Treatments
+{
+private:
+    std::vector<AbstractTreatment<IntegerRaster, FloatRaster>*> treatments;
+public:
+    ~Treatments()
+    {
+        for (auto item : treatments)
+        {
+            delete item;
         }
     }
-    void apply_treatment_host(int year, IntegerRaster &infected, IntegerRaster &susceptible)
+    /*!
+     * \brief Add treatment, based on parameters it is distinguished
+     * which treatment it will be.
+     *
+     * \param map treatment raster
+     * \param start_date date when treatment is applied
+     * \param num_days for simple treatments should be 0, otherwise number of days host is resistant
+     * \param treatment_application if efficiency < 100% how should it be applied to infected/susceptible
+     * \param increase_by_step function to increase simulation step
+     */
+    void add_treatment(const FloatRaster& map, const Date& start_date, int num_days, TreatmentApplication treatment_application,
+                       std::function<void (Date&)> increase_by_step)
     {
-        // this expression fails in rcpp
-        // host = host - (host * treatments[year]);
-        if (treatments.find(year) != treatments.end()) {
-            for(unsigned i = 0; i < infected.rows(); i++)
-                for(unsigned j = 0; j < infected.cols(); j++) {
-                    if (application == TreatmentApplication::Ratio) {
-                        infected(i, j) = infected(i, j) - (infected(i, j) * treatments[year](i, j));
-                        susceptible(i, j) = susceptible(i, j) - (susceptible(i, j) * treatments[year](i, j));
-                    }
-                    else if (application == TreatmentApplication::AllInfectedInCell) {
-                        infected(i, j) = treatments[year](i, j) ? 0 : infected(i, j);
-                        susceptible(i, j) = susceptible(i, j) - (susceptible(i, j) * treatments[year](i, j));
-                    }
-                }
+        if (num_days == 0)
+            treatments.push_back(new SimpleTreatment<IntegerRaster, FloatRaster>(map, start_date, treatment_application, increase_by_step));
+        else
+            treatments.push_back(new PesticideTreatment<IntegerRaster, FloatRaster>(map, start_date, num_days, treatment_application, increase_by_step));
+    }
+    /*!
+     * \brief Do management if needed.
+     * Should be called before every simulation step.
+     * Decides internally whether any treatment needs to be
+     * activated/deactivated.
+     *
+     * \param current simulation date
+     * \param infected raster of infected host
+     * \param susceptible raster of susceptible host
+     * \param resistant raster of resistant host
+     * \return true if any management action was necessary
+     */
+    bool manage(const Date& current, IntegerRaster& infected,
+                IntegerRaster& susceptible, IntegerRaster& resistant)
+    {
+        bool applied = false;
```

Comment:
I actually really liked the way it was listed previously because the function is applying the management while changing the underlying states of the pools in the population. Unless you are thinking about tracking the state of the management but even then as applied makes sense because if True then it has been applied if False it has not been. This makes much more sense from a management perspective IMO.

Reply:
The function returns true 1) if treatment was applied (with apply_treatment()) and 2) also if the treatment finished/expired/ended (with end_treatment()). I just think it is confusing that the internal variable (`applied`) and return value documentation ("true if any management action was necessary") refer to actions and applications which at least seem to refer to beginning of the treatment rather than the end.

Perhaps the question is if the function should indicate "treatment state changed" (current situation) versus "treatment was applied". Alternative would be to indicate the change in finer manner using an enum or, in case we can't decide, indicate nothing if the return value is now unused (@petrasovaa ?).

Reply:
Good point on the true if ended as well. 
So are you thinking something like:
Treatment_state: 'planned'
Treatment_state: 'applied'
Treatment_state: 'ended'

This might be nice if we are to be able to pass them out in a steering context which means there might be some changes needed if that were the case ( e.g. a treatment is applied but not ended between steering events, might not be a problem for true steering but would be for OI (output -> input)  based steering. 

Thoughts on possible solutions or a good way to be able to handle that in those situations.

Reply:
Some clarification: the return status ('applied') was there to optimize calling of the 'manage' method for other stochastic runs. Since  'manage' is called every time step for each of the stochastic runs, I thought it may be good to return if any management action was needed for the first run and if not, skip calling it for other runs. This is not usable when we use parallel implementation with openMP in the grass version.

Regarding steering, as long as you export also resistant raster, it should work as it is now, at least it was intended to work... Let me know if more clarification is needed.